File: exercises/day14-Santa-package-factory-exercises/typescript/src/packagingservice/PackagingService.ts
import { Gift } from "./Gift";
import { Child } from "./Child";
import { GiftSize, PackageType } from "./enums";

export class PackagingService {
  determinePackageType(gift: Gift, child: Child): PackageType {
    if (gift.size === GiftSize.EXTRA_LARGE) {
      return PackageType.SPECIAL_CONTAINER;
    }

    if (gift.isFragile) {
      return gift.size === GiftSize.SMALL
        ? PackageType.BOX_SMALL
        : PackageType.BOX_MEDIUM;
    }

    if (child.age < 5) {
      return PackageType.GIFT_BAG;
    }

    switch (gift.size) {
      case GiftSize.SMALL:
        return PackageType.BOX_SMALL;
      case GiftSize.MEDIUM:
        return PackageType.BOX_MEDIUM;
      case GiftSize.LARGE:
        return PackageType.BOX_LARGE;
      default:
        return PackageType.WRAPPER_PAPER;
    }
  }

  canPackageGift(gift: Gift, child: Child): boolean {
    if (!child.hasBeenNice) {
      return false;
    }

    if (gift.recommendedMinAge > child.age) {
      return false;
    }

    return true;
  }
}

File: exercises/day14-Santa-package-factory-exercises/scala/src/main/scala/packagingservice/PackagingService.scala
package packagingservice

import packagingservice.GiftSize._
import packagingservice.PackageType._

class PackagingService {

  def determinePackageType(gift: Gift, child: Child): PackageType = {
    gift.size match {
      case EXTRA_LARGE =>
        SPECIAL_CONTAINER

      case _ if gift.isFragile =>
        gift.size match {
          case SMALL => BOX_SMALL
          case _     => BOX_MEDIUM
        }

      case _ if child.age < 5 =>
        GIFT_BAG

      case SMALL  => BOX_SMALL
      case MEDIUM => BOX_MEDIUM
      case LARGE  => BOX_LARGE
      case _      => WRAPPER_PAPER
    }
  }

  def canPackageGift(gift: Gift, child: Child): Boolean = {
    if (!child.hasBeenNice) {
      false
    } else if (gift.recommendedMinAge > child.age) {
      false
    } else {
      true
    }
  }
}
